### ISSUE TITLE
Invalid DDOs can be published on-chain, and Aquarius won't index them

### ISSUE THREAD

**Issue by trentmc** [oceanprotocol/market#367]
**Describe the bug**
In Rinkeby, asset doesn't show up in Ocean Market after publishing. Can't set price etc.

**To Reproduce**
1. Open browser.
2. Have Metamask on, with wallet connected to Rinkeby. Make sure the account has enough ETH.
3. Go to market.oceanprotocol.com. 
4. Click 'connect wallet' on top right
5. Click on 'PUBLISH'
6. Fill out the whole form with test info; check the "I agree to these Terms and Conditions" box. Name of asset is "branin2".
7. Click 'SUBMIT' button
8. The page will say "Publishing Data Set : 1/5 .." then "2/5".
9. Metamask will pop up; click confirm. Metamask will pop up again; click confirm.
10.  Page will say "Publishing Data Set : 3/5". Then "Successfully published. Now create a price on your dataset". It has a button "Go to dataset"
11. Click on the button "Go to dataset". 
12. It goes to a page that says "Could not retrieve asset. The DDO for did:op:297E5671E9EB802b3E4494f8BeF1d84209161dea was not found in MetadataCache. If you just published a new data set, wait some seconds and refresh this page."
13. So I waited a minute, and refreshed. Go same error message
14. So I waited 15 min, and refreshed. Still the same error message
15. I opened a separate browser tab, went to market.oceanprotocol.com, and scrolled down to "New data sets" section. The asset ("Branin2") is not there. Nothing's there in the last hour, so it's easy to tell.

**Expected behavior**
To be able to publish properly, including:
- Seeing the asset in Ocean Market
- Setting a fixed price or doing AMM pool

**Desktop (please complete the following information):**
 - OS: Win 10
 - Browser: Firefox

**Screenshots**

Ocean Market 1
![publish1](https://user-images.githubusercontent.com/5305452/107145922-2c30fd00-6945-11eb-968b-cdba12f972dc.png)

Ocean Market 2
![publish2](https://user-images.githubusercontent.com/5305452/107145968-7e721e00-6945-11eb-9e86-d246d82468b4.png)

Ocean Market 3
![publish3](https://user-images.githubusercontent.com/5305452/107145969-80d47800-6945-11eb-91b3-995af6d2b43c.png)

Ocean Market 4 - see the DDO error
![publish4 - ddo error](https://user-images.githubusercontent.com/5305452/107145974-829e3b80-6945-11eb-962f-ecc675b3cf8f.png)

Ocean Market 5 - DDO error after waiting a few min
![publish4 - ddo error after waiting](https://user-images.githubusercontent.com/5305452/107145978-85992c00-6945-11eb-8625-a80def5c4cd5.png)

Ocean Market 6 - DDO error after waiting many min
![publish4 - ddo error after waiting 15 min](https://user-images.githubusercontent.com/5305452/107145984-9053c100-6945-11eb-8dc6-98b1b18393b1.png)

Ocean Market - not shown in most recent assets
![publish6 - data asset not in ocean market new data sets](https://user-images.githubusercontent.com/5305452/107145987-98abfc00-6945-11eb-8f3e-e5fa1da4bde5.png)

Metamask txs
![metamask txs](https://user-images.githubusercontent.com/5305452/107145991-9c3f8300-6945-11eb-91d2-cb37e3dca0e9.png)

Etherscan txs. For [this address](https://rinkeby.etherscan.io/address/0x15f8a84B184A62bD3fC5Ad575F85560C72BAFB24), got [tx 1](https://rinkeby.etherscan.io/tx/0x57c87ba35a90abdf62f890004308d7ddcd562828375b98ba42f729164fb595e0) and [tx 2](https://rinkeby.etherscan.io/tx/0x79fee6ad519347cc9ad7d6a9d24d329a6292f3a05eaf3b1fad7168847d517851)

**Comment by kremalicious** [comment]
No clue how to reproduce. Seems like an Aquarius issue, in the last weeks it went down on Rinkeby almost daily. Asset simply does not exist there: https://aquarius.rinkeby.oceanprotocol.com/api/v1/aquarius/assets/ddo/did:op:297E5671E9EB802b3E4494f8BeF1d84209161dea

Multiple sets were published on Rinkeby in last hours:
![Screen Shot 2021-02-07 at 18.32.10.png](https://images.zenhubusercontent.com/5e1f103b5b295ab1a54f1e87/b70315ec-5c91-462d-8d03-170a0338c8ca)

**Comment by trentmc** [comment]
Did you try reproducing with the steps given? It happened to me two times in a row. 

And if it is Aquarius issues:
- there is no ux guidance on what's happening or what the user can do 
- the market is not able to handle the Aquarius errors robustly, eg try again or give a way to try the remaining steps later

**Comment by kremalicious** [comment]
agree, user-friendly error handling is terrible but this is also on Aquarius side resulting in `market` having no reliable way of detecting specific errors. This is a shortcoming of its API implementation described a bit here https://github.com/oceanprotocol/aquarius/issues/273

**Comment by trentmc** [comment]
Thinking out loud, if we assume that it is Aquarius causing the market ux issues, then the easy way to reproduce it is to turn off Aquarius / make it act badly. 

Then the Market ux can be improved to give the user better guidance (rather than an error message saying "please wait a few seconds and refresh").

**Comment by kremalicious** [comment]
As of right now we cannot differentiate if Aquarius is down or an asset is not found. All is a `500` from the `market` perspective cause that's what Aquarius tells us. If publishing fails somewhere in between, the UI shows a retry button but only based on transactions happening. After that we have to rely on Aquarius which we kinda can't.

As I said, I cannot reproduce, just tried another test asset and all goes successful for me on Rinkeby https://market.oceanprotocol.com/asset/did:op:94590D392B91EA6Fbc166B69de705e989e1ED33d

**Comment by trentmc** [comment]
Here's perhaps the most robust solution, though it will take more work:

Don't use Aquarius to publish. 

After all, it's only supposed to be a metadata cache. But when you're publishing you don't need a cache. You just have to get some data into ddo.sol contract.

This would likely require writing a good chunk of JS. But it would be far more robust. It also sets up nicely for v4 which will put metadata into the erc20 itself.

**Comment by kremalicious** [comment]
We use Aquarius to display an asset and there is no other way right now. As I said, nothing in the publish fails from the UI side, but when retrieving a single asset it fails. There is no way of retrieving a single asset with its metadata from somewhere else.

**Comment by kremalicious** [comment]
code example to show that Aquarius is already not used when publishing: https://github.com/oceanprotocol/ocean.js/blob/main/src/ocean/Assets.ts#L192-L199

So ocean.js tells us DDO is now on-chain, then we try to retrieve it cause we want to publish it but Aquarius does not have it. No way of detecting or reacting to that on the market.

**Comment by kremalicious** [comment]
I was able to reproduce this locally with #369 and this is the error Aquarius show in its logs:

```
aquarius_1         | 2021-02-09 17:03:52,105 - aquarius.events.events_monitor - ERROR - New ddo has validation errors: [{'path': 'main/files/0/contentLength', 'message': "None is not of type 'string'"}]
```

So in my case Aquarius wants a `contentLength` on the files but provider did not detect it. This is a new bug and was not present before we switched to using the provider as a file checker 5 days ago in #352

**Comment by kremalicious** [comment]
#91 fixes the root cause of this issue. We still need DDO validation before form submission to prevent users from running into similar issues for which we need to wait for resolve of https://github.com/oceanprotocol/ocean.js/issues/400

**Comment by trentmc** [comment]
Good job working to get to the root of this!

I think I supplied this:
https://github.com/trentmc/branin/branin.arff

I did not check then (my bad). I see now that it returns a "not found" message. 

It's also possible that I supplied the proper url:
https://github.com/trentmc/branin/blob/master/branin.arff

**Comment by bogdanfazakas** [comment]
Should we set a default value from the market if the `contentLength` is not received from provider when checking fileinfo ? We do the same think here https://github.com/oceanprotocol/market/blob/main/src/utils/provider.ts#L31-L32 with the `contentType ` but that's a required parameter in the FIle interface.

**Comment by kremalicious** [comment]
yeah think so, then we're really safe and it is no overhead at all. Could you prepare a PR for it?

On principle, we should always only reflect what Aquarius' metadata validation wants in the library typings, and then in the clients. But hey, tech reality!

**Comment by bogdanfazakas** [comment]
sure , I'll handle this.

**Comment by kremalicious** [comment]
we will have this closed by #372, which will prevent the reason for the original issue from happening again, in tandem with changes on `provider`.

At same time there are multiple other ways which would further help users in similar cases of "faulty" metadata. These ways are outlined in #373 and #366 and will be tackled individually there
